flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/triggers/AsyncTrigger.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.asyncprocessing.operators.windowing.triggers;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.api.common.functions.RuntimeContext;
+import org.apache.flink.api.common.state.v2.State;
+import org.apache.flink.api.common.state.v2.StateDescriptor;
+import org.apache.flink.api.common.state.v2.StateFuture;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.streaming.api.windowing.triggers.TriggerResult;
+import org.apache.flink.streaming.api.windowing.windows.Window;
+
+import java.io.Serializable;
+
+/**
+ * A {@code Trigger} determines when a pane of a window should be evaluated to emit the results for
+ * that part of the window. This trigger is for the async window operator, i.e., {@link
+ * org.apache.flink.runtime.asyncprocessing.operators.windowing.AsyncWindowOperator}.
+ *
+ * <p>All State APIs in the async trigger are from State V2.
+ *
+ * <p>A pane is the bucket of elements that have the same key (assigned by the {@link
+ * org.apache.flink.api.java.functions.KeySelector}) and same {@link Window}. An element can be in
+ * multiple panes if it was assigned to multiple windows by the {@link
+ * org.apache.flink.streaming.api.windowing.assigners.WindowAssigner}. These panes all have their
+ * own instance of the {@code Trigger}.
+ *
+ * <p>Triggers must not maintain state internally since they can be re-created or reused for
+ * different keys. All necessary state should be persisted using the state abstraction available on
+ * the {@link TriggerContext}.
+ *
+ * <p>When used with a {@link
+ * org.apache.flink.streaming.api.windowing.assigners.MergingWindowAssigner} the {@code Trigger}
+ * must return {@code true} from {@link #canMerge()} and {@link #onMerge(Window, OnMergeContext)}
+ * most be properly implemented.
+ *
+ * @param <T> The type of elements on which this {@code Trigger} works.
+ * @param <W> The type of {@link Window Windows} on which this {@code Trigger} can operate.
+ */
+@Experimental
```

Comment:
We'd better mark all the async triggers as `@Internal` since they are not exposed to users.

Reply:
resolved